### ISSUE TITLE
Break out fromInteger/fromRational from Num/Fractional?

### ISSUE THREAD

**Issue by brandonchinn178** [ghc-proposals/ghc-proposals#438]
There have been a few cases where I would like to write a numeric literal in place of a custom data type, but that data type does not have valid implementations for the other functions in the type class (or you don't want to implement the other functions for various reasons). If I were redesigning everything from scratch, I'd do:
```hs
-- integral literals
class FromInteger a where
  fromInteger :: Integer -> a

-- decimal literals
class FromRational a where
  fromRational :: Rational -> a

-- string literals, renamed from IsString
class FromString s where
  fromString :: String -> s

-- new IsString; I've always wanted this anyway,
-- to generalize functions that can take in any
-- string-like thing (e.g. logging or golden tests)
class FromString s => IsString s where
  toString :: s -> String

-- list literals, broken out from IsList
class FromList l where
  type Item l
  fromList :: [Item l] -> l

-- new IsList
class FromList l => IsList l where
  toList :: l -> [Item l]
```

For example, we have a `BigDecimal` newtype over `Scientific` because `Scientific` has unsafe instances. This way, we can provide explicit `unsafeAdd` functions and prevent people from accidentally using `+`. But currently, there's no way to use a literal `10` to be `BigDecimal 10`. With this proposal, one could implement `FromInteger` but not `Num` as a whole.

This is mentioned as part of https://github.com/ghc-proposals/ghc-proposals/pull/124, but I think it's a worthwhile unit separate from the "prevent `fromInteger` from being partial" discussion. Obviously, this is a big breaking change, but I just wanted to write this up and see if others have any thoughts on this before trying to come up with a proposal/migration plan.

**Comment by phadej** [comment]
I would love something like this.

---

```
class FromNatural a where
  fromNatural :: Natural -> a
```

would be useful, it's more constrained, and we need `Integer` only for `NegativeLiterals`.

---

Some will definitely mention that "can't we make it so `257 :: Word8` would fail at compile time, without compiler support. And similarly `"\1000" :: ByteString`. There are discussions about that issue, and some ideas too. A proposal would need to articulate well why you do something simple(r).

---

Note that `toList` is needed to desugar patterns.

```haskell
{-# LANGUAGE OverloadedLists #-}
unsingleton :: Map k v -> [(k, v)]
unsingleton [kv] = Just kv
unsingleton _ = Nothing
```

Strings and number literal pattern, they then require `Eq`.

**Comment by brandonchinn178** [comment]
Related: #451

**Comment by brandonchinn178** [comment]
It would also be cool to generalize all overloaded literal functions, like
```hs
class Lift b => FromLiteral a b where
    fromLiteral :: a -> Either String b
```
where you could return `Left` with an error message, which would be realized as a compile time error. This would allow us to create a partial `fromInteger` for `Natural`, a guarantee that `NonEmpty` is total at runtime, etc.

Related: #482

**Comment by nomeata** [comment]
I think these days the Core Libraries Committee proposal project (https://github.com/haskell/core-libraries-committee/) would be the right place to discuss such library changes.

**Comment by brandonchinn178** [comment]
I dont think so; this requires changes to how the compiler treats literal numbers, strings, and lists. e.g. instead of OverloadedStrings doing the conversion
```hs
"hi"
-- to
fromString "hi"
-- using (IsString Text)
```
it would do
```hs
"hi"
-- to
$(either fail lift $ fromLiteral "hi")
-- using (FromLiteral String Text)
```

**Comment by nomeata** [comment]
Yes, I noticed and deleted my comment, sorry for the noise.

**Comment by noughtmare** [comment]
> where you could return Left with an error message, which would be realized as a compile time error. This would allow us to create a partial fromInteger for Natural, a guarantee that NonEmpty is total at runtime, etc.

I don't think this would be possible as it requires arbitrary compile time evaluation. What if the implementation is an infinite loop?

**Comment by brandonchinn178** [comment]
I don't see why it'd be any different than TH. If it's an infinite loop, then it will hang at compile -time. But it's definitely possible

**Comment by noughtmare** [comment]
The difference is that TH is opt in and very explicitly warned about. This is one of the reasons most people should avoid writing nontrivial TH code. Perhaps it would be acceptable if the type explicitly mentions TH, like have a separate method:

```haskell
class FromLiteral a b where
    fromLiteral :: a -> b
    fromLiteralTH :: Quasi m => Code m a -> Code m b
    fromLiteralTH x = [|| fromLiteral $$x ||]
```

And then do the desugaring like this:

```
"hi"
-- to
$$(fromLiteralTH [|| "hi" ||])
-- using (FromLiteral String Text)
```
